### PR TITLE
Add support for unwrapping of parameters, and determining trainability

### PR DESCRIPTION
**Pull request by josh146**
**Context:** Cleans up the behaviour of `qml.math`, and adds better support for (a) unwrapping lists of tensors to NumPy arrays, and (b) determining trainability.

**Description of the Change:**

- `qml.math.requires_grad` is updated to work properly for both Autograd and JAX parameters, on both the forward and backward passes.

- `qml.math.get_trainable_indices` is added, which allows a sequence of tensor-like objects to be passed, and the indices corresponding to trainable tensors being returned.

- `qml.math.unwrap` unwraps a sequence of tensor-like objects to NumPy arrays.

**Benefits:** Clean and well-tested, interface agnostic methods for determining tensor trainability and conversions to NumPy arrays.

**Possible Drawbacks:** For the Autograd interface, assuming that `requires_grad=True` on non-tensor objects remains hard-coded in (for now).

**Related GitHub Issues:** n/a

### CODE REVIEW

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -406,6 +406,19 @@
 * Added the `id` attribute to templates, which was missing from 
   PR [(#1377)](https://github.com/PennyLaneAI/pennylane/pull/1377).
   [(#1438)](https://github.com/PennyLaneAI/pennylane/pull/1438)
+
+* The `qml.math` module, for autodiff framework-agnostic tensor manipulation,
```
Comment (mariaschuld) [inline .github/CHANGELOG.md:410]:
```suggestion
* The `qml.math` module for autodiff framework-agnostic tensor manipulation,
```

Comment (mariaschuld) [reply .github/CHANGELOG.md:410]: Or do I sense an upcoming move away from the word "interface"? :)

File: pennylane/math/multi_dispatch.py [hunk 2]
```diff
@@ -250,6 +252,38 @@ def dot(tensor1, tensor2):
...
+def get_trainable_indices(values):
+    """Returns a set containing the trainable indices of a sequence of
+    values.
+
+    Args:
+        values (Sequence[tensor_like]): Sequence of tensor-like objects to inspect
+
+    Returns:
+        set[int]: set containing the indices of the trainable tensor-like objects
+        within the input sequence.
```
Comment (mariaschuld) [inline pennylane/math/multi_dispatch.py:264]:
```suggestion
        set[int]: Set containing the indices of the trainable tensor-like objects
        within the input sequence.
```

File: pennylane/math/multi_dispatch.py [hunk 3]
```diff
@@ -250,6 +252,38 @@ def dot(tensor1, tensor2):
     return np.dot(x, y, like=interface)
 
 
+def get_trainable_indices(values):
+    """Returns a set containing the trainable indices of a sequence of
+    values.
+
+    Args:
+        values (Sequence[tensor_like]): Sequence of tensor-like objects to inspect
```
Comment (mariaschuld) [inline pennylane/math/multi_dispatch.py:260]: If I [understand correctly](https://docs.python.org/3/library/collections.abc.html), you only need an `Iterable` here, since `enumerate` does not need methods like length or indexing, right?

File: pennylane/math/multi_dispatch.py [hunk 4]
```diff
@@ -250,6 +252,38 @@ def dot(tensor1, tensor2):
...
+    tensor(0.0899685, requires_grad=True)
+    """
+    interface = _multi_dispatch(values)
+    trainable_params = set()
+
+    for idx, p in enumerate(values):
+        if requires_grad(p, interface=interface):
+            trainable_params.add(idx)
+
+    return trainable_params
```
Comment (mariaschuld) [inline pennylane/math/multi_dispatch.py:284]: :100:

File: pennylane/math/multi_dispatch.py [hunk 5]
```diff
@@ -340,3 +374,50 @@ def where(condition, x, y):
     tensor([ 0.6000,  0.2300,  0.7000, -4.0000, -5.0000], grad_fn=<SWhereBackward>)
     """
     return np.where(condition, x, y, like=_multi_dispatch([condition, x, y]))
+
+
+def unwrap(values, max_depth=None):
+    """Unwrap a sequence of objects to NumPy arrays.
+
+    Note that tensors on GPUs will automatically be copied
+    to the CPU.
```
Comment (mariaschuld) [inline pennylane/math/multi_dispatch.py:383]: I wouldn't have thought of this, nice to be aware!

File: pennylane/math/multi_dispatch.py [hunk 6]
```diff
@@ -340,3 +374,50 @@ def where(condition, x, y):
...
+
+    >>> def cost_fn(params):
+    ...     unwrapped_params = math.unwrap(params)
+    ...     print("Unwrapped:", [(i, type(i)) for i in unwrapped_params])
+    ...     return np.sum(np.sin(params))
+    >>> params = np.array([0.1, 0.2, 0.3])
+    >>> grad = autograd.grad(cost_fn)(params)
+    Unwrapped: [(0.1, <class 'float'>), (0.2, <class 'float'>), (0.3, <class 'float'>)]
+    >>> print(grad)
+    [0.99500417 0.98006658 0.95533649]
```
Comment (mariaschuld) [inline pennylane/math/multi_dispatch.py:407]: Nice!

File: pennylane/math/multi_dispatch.py [hunk 7]
```diff
@@ -340,3 +374,50 @@ def where(condition, x, y):
...
+    res = []
+
+    for t in values:
+        if isinstance(t, ArrayBox):
+            a = np.to_numpy(t, max_depth=max_depth)
+        else:
+            a = np.to_numpy(t)
+
+        if isinstance(a, ndarray) and not a.shape:
+            # if NumPy array is scalar, convert to a Python float
```
Comment (mariaschuld) [inline pennylane/math/multi_dispatch.py:418]: what if we have something like `tf.Variable(0.1)` ? Would you also expect a float? In other words, do we always "squeeze" scalars here?

File: pennylane/math/utils.py [hunk 8]
```diff
@@ -193,6 +194,8 @@ def requires_grad(tensor):
 
     Args:
         tensor (tensor_like): input tensor
+        interface (str): The name of the interface. If not provided,
+            will be determined automatically.
```
Comment (mariaschuld) [inline pennylane/math/utils.py:198]:
```suggestion
        interface (str): The name of the interface. Will be determined automatically 
        if not provided.
```

Splitting hairs here :)

File: pennylane/math/utils.py [hunk 9]
```diff
@@ -229,7 +232,7 @@ def requires_grad(tensor):
     ...     print(requires_grad(x))
     True
     """
-    interface = get_interface(tensor)
+    interface = interface or get_interface(tensor)
```
Comment (mariaschuld) [inline pennylane/math/utils.py:235]: Can't we always just do this? Or do you want to save costs if we already know the interface?

File: pennylane/math/utils.py [hunk 10]
```diff
@@ -241,13 +244,23 @@ def requires_grad(tensor):
...
+
+    if interface == "torch":
+        return getattr(tensor, "requires_grad", False)
 
     if interface == "numpy":
         return False
 
     if interface == "jax":
-        return True
+        import jax
```
Comment (mariaschuld) [inline pennylane/math/utils.py:262]: There is probably no other way, but won't code reload the library every time we build a tape here?

File: pennylane/math/utils.py [hunk 11]
```diff
@@ -241,13 +244,23 @@ def requires_grad(tensor):
...
+
+        return isinstance(tensor, jax.interpreters.ad.JVPTracer)
```
Comment (mariaschuld) [inline pennylane/math/utils.py:264]: nice! Is this a hack or intended way?

File: tests/math/test_functions.py [hunk 12]
```diff
@@ -661,10 +662,26 @@ def test_numpy(self, t):
...
-    @pytest.mark.parametrize("t", [jnp.array([1, 2, 3])])
-    def test_jax(self, t):
-        """jax.DeviceArrays will always return True"""
-        assert fn.requires_grad(t)
+    def test_jax(self):
+        """JAX DeviceArrays differentiability depends on the argnums argument"""
+        res = None
+
+        def cost_fn(t, s):
+            nonlocal res
```
Comment (mariaschuld) [inline tests/math/test_functions.py:670]: That is so clever, I never knew that trick to look into the function!

File: tests/math/test_functions.py [hunk 13]
```diff
@@ -674,6 +691,38 @@ def test_autograd(self):
         t = np.array([1.0, 2.0], requires_grad=False)
         assert not fn.requires_grad(t)
 
+    def test_autograd_backwards(self):
+        """jax.DeviceArrays will always return True"""
```
Comment (mariaschuld) [inline tests/math/test_functions.py:695]: maybe a bit more info from the docstring would help? Like "will always return True in a backwards pass"(?)

Comment (mariaschuld) [reply tests/math/test_functions.py:695]: wait, this is a typo right? the test is not about jax :)

File: tests/math/test_functions.py [hunk 14]
```diff
@@ -1392,3 +1441,178 @@ def test_torch_coercion(self):
...
+            return np.sum(np.sin(params[0] * params[2])) + params[1]
+
+        values = [onp.array([0.1, 0.2]), np.tensor(0.1, dtype=np.float64), np.tensor([0.5, 0.2])]
+        cost_fn(values)
+
+        expected = [np.array([0.1, 0.2]), 0.1, np.array([0.5, 0.2])]
+        assert all(np.allclose(a, b) for a, b in zip(unwrapped_params, expected))
+        assert all(not isinstance(a, np.tensor) for a in unwrapped_params)
+
+    def test_autograd_unwrapping_backward(self):
```
Comment (mariaschuld) [inline tests/math/test_functions.py:1488]: just curious, why does autograd need backward and forward testing, but not the other interfaces?

File: tests/math/test_functions.py [hunk 15]
```diff
@@ -1392,3 +1441,178 @@ def test_torch_coercion(self):
...
+        values = [jnp.array([0.1, 0.2]), onp.array(0.1, dtype=np.float64), jnp.array([0.5, 0.2])]
+        cost_fn(values)
+
+        expected = [np.array([0.1, 0.2]), 0.1, np.array([0.5, 0.2])]
+        assert all(np.allclose(a, b) for a, b in zip(unwrapped_params, expected))
+        assert all(not isinstance(a, np.tensor) for a in unwrapped_params)
+
+
+class TestGetTrainable:
+    """Test getting trainable indices"""
```
Comment (mariaschuld) [inline tests/math/test_functions.py:1546]:
```suggestion
    """Tests for getting trainable indices"""
```
Maybe?

File: tests/math/test_functions.py [hunk 16]
```diff
@@ -1392,3 +1441,178 @@ def test_torch_coercion(self):
...
+    def test_tensorflow(self):
+        """Test that the trainability indices of a sequence of TensorFlow values
+        is correctly extracted"""
+        values = [
+            onp.array([0.1, 0.2]),
+            tf.Variable(0.1, dtype=tf.float64),
+            tf.constant([0.5, 0.2]),
+        ]
+
+        # outside of a gradient tape, no indices are trainable
```
Comment (mariaschuld) [inline tests/math/test_functions.py:1557]: Wow, that's also clever, haven't thought about it

**Review by mariaschuld — APPROVED** [review]
Only one serious typo to fix in the test, the remainder are just questions. But this is a serious improvement!

Nice test case by the way, with the nonlocal variable spy...

**Comment by josh146** [comment]
[ch7903]

File: .github/CHANGELOG.md (hunk 1 above)
Comment (josh146) [reply .github/CHANGELOG.md:414]: > Or do I sense an upcoming move away from the word "interface"? :)

😆 This is something I realized from the ADR. I'm trying to move towards a place where:

- `interface`: the specific code logic in PL for gluing together autodiff frameworks and quantum gradients
- `autodiff framework/library`: Torch, TF, etc.

Currently, `interface` means too many things, so I think it makes sense to narrow the scope.

Comment (josh146) [reply .github/CHANGELOG.md:414]:
```suggestion
* The `qml.math` module, for framework-agnostic tensor manipulation,
```

File: pennylane/math/multi_dispatch.py (hunk 3 above)
Comment (josh146) [reply pennylane/math/multi_dispatch.py:260]: Oh nice catch!
```suggestion
        values (Iterable[tensor_like]): Sequence of tensor-like objects to inspect
```

File: pennylane/math/multi_dispatch.py (hunk 7 above)
Comment (josh146) [reply pennylane/math/multi_dispatch.py:418]: We do. This is an issue we have with some plugins.

Even though `default.qubit` treats `np.array(0.1)` the same as `0.1`, not all QC frameworks do. In particular, Qiskit and Braket have logic that will fail if given a scalar array!

c.f. https://github.com/PennyLaneAI/pennylane/pull/893

File: pennylane/math/utils.py (hunk 8 above)
Comment (josh146) [reply pennylane/math/utils.py:198]: Yours reads better!

File: pennylane/math/multi_dispatch.py (hunk 5 above)
Comment (josh146) [reply pennylane/math/multi_dispatch.py:383]: I can't claim the credit for this unfortunately, `autoray.to_numpy()` has this built in 😆 

A nice artifact of using `autoray`, since I don't have a GPU and wouldn't have been able to test this myself.

File: .github/CHANGELOG.md (hunk 1 above)
Comment (mariaschuld) [reply .github/CHANGELOG.md]: actually, feel free to leave the original then, it is more explicit!

File: pennylane/math/utils.py (hunk 9 above)
Comment (josh146) [reply pennylane/math/utils.py:235]: Oh, actually this is to enable different behaviour! Sometimes, we want to consider a singular interface for an entire sequence of tensors, rather than on a per-tensor basis.

E.g., imagine if we had a iterable of the form

```python
[autograd, autograd, tf, tf, autograd]
```

which could feasibly happen if a user is using PennyLane NumPy for constant, non-differentiable data. However, at the moment we assume `requires_grad=True` by default for PennyLane NumPy! So if we treat each element individually, we would get this:

```python
[True, True, True, True, True]
```

Instead, if we pass `interface="tf"`, then we tell this function to 'use the TF method of determining trainability, regardless of the tensor type'. As a result, we get the expected output:

```python
[False, False, True, True, False]
```

File: pennylane/math/utils.py (hunk 11 above)
Comment (josh146) [reply pennylane/math/utils.py:264]: Not so much a hack as 'undocumented behaviour'. Since it is not a public API, JAX could easily change it without warning in a future version.

File: pennylane/math/utils.py (hunk 10 above)
Comment (josh146) [reply pennylane/math/utils.py:262]: Nope! Python will only import the library the first time the import statement is called, so there is no import overhead (https://stackoverflow.com/questions/32151193/is-there-a-performance-cost-putting-python-imports-inside-functions)

File: tests/math/test_functions.py (hunk 13 above)
Comment (josh146) [reply tests/math/test_functions.py:695]: Oops 🤦 
```suggestion
        """Autograd trainability corresponds to the requires_grad attribute during the backwards pass."""
```

File: tests/math/test_functions.py (hunk 14 above)
Comment (josh146) [reply tests/math/test_functions.py:1488]: Because Autograd behaves differently during the backward and forward passes :(

- On the forward pass, it simply passes `np.tensor` objects. They are unwrapped by calling `.numpy()`.

- On the backward pass, there is an implicit forward pass where it instead passes `ArrayBox` objects. They are unwrapped by calling `._value`.

There is logic in `qml.math` to account for both cases, and I want to ensure it works correctly.

TF and Torch are different:

- they do not have a functional user-interface, so the forward pass and the backward pass always occur separately
- As a result, the forward pass is always explicit, and always just receives a tensor.

File: tests/math/test_functions.py (hunk 16 above)
Comment (josh146) [reply tests/math/test_functions.py:1557]: It is new behaviour in TF 2.3+ 😆 In earlier versions, `Variable` objects were always trainable, and `constants` never trainable, regardless of a tape being present or not.
